### PR TITLE
Implement the batching functionality on history event replication task

### PR DESCRIPTION
**Pull request by xwduan**
**What changed?**
Implement the batching functionality on history event replication task

**Why?**
Improve performance of history event replication task

**How did you test it?**
manually tested locally. Unit tests.

**Potential risks**
N/a. Feature is behind a feature flag.

**Is hotfix candidate?**
no

### CODE REVIEW

File: service/history/replication/fx.go [hunk 1]
```diff
@@ -119,18 +120,32 @@ func ReplicationTaskExecutorProvider() TaskExecutorProvider {
...
 	)
 }
 
+func SequentialTaskQueueFactoryProvider(
+	logger log.Logger,
+	metricsHandler metrics.Handler,
+	config *configs.Config,
+) ctasks.SequentialTaskQueueFactory[TrackableExecutableTask] {
+	return func(task TrackableExecutableTask) ctasks.SequentialTaskQueue[TrackableExecutableTask] {
+		if config.EnableReplicationTaskBatching() {
```
Comment (xwduan) [inline service/history/replication/fx.go:142]: Feature flag here maybe not enough? If the task for same workflow keep coming, then no new queue will be created, and we need to restart history pod to pick up the change.

File: service/history/replication/sequential_batch_queue.go [hunk 2]
```diff
@@ -41,8 +41,9 @@ type (
 		taskQueue                    collection.Queue[*batchedTask]
 		lastTask                     *batchedTask
 		batchedIndividualTaskHandler func(task TrackableExecutableTask)
```
Comment (wxing1292) [inline service/history/replication/sequential_batch_queue.go:43]: this var is not used anywhere anymore?

File: service/history/replication/executable_history_task.go [hunk 3]
```diff
@@ -297,3 +302,145 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
 	}
 	return events, newRunEvents, err
 }
+
+func (e *ExecutableHistoryTask) BatchWith(incomingTask BatchableTask) (_ TrackableExecutableTask, retErr error) {
```
Comment (wxing1292) [inline service/history/replication/executable_history_task.go:306]: `func (e *ExecutableHistoryTask) BatchWith(incomingTask BatchableTask) (TrackableExecutableTask, bool) {`?

return a bool indicating whether batch is successful is more appropriate

Comment (wxing1292) [reply service/history/replication/executable_history_task.go:306]: i see, validation is also done internally .....

File: service/history/replication/executable_history_task.go [hunk 4]
```diff
@@ -297,3 +302,145 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
...
+	currentLastIndex := len(e.versionHistoryItems) - 1
+	currentLastItem := e.versionHistoryItems[currentLastIndex]
+	if firstNotEqualIndex == len(e.versionHistoryItems) || // current version history is part of incoming version history
+		(firstNotEqualIndex == currentLastIndex &&
+			currentLastItem.Version == incomingHistoryItems[firstNotEqualIndex].Version &&
+			currentLastItem.EventId <= incomingHistoryItems[firstNotEqualIndex].EventId) {
+		return nil
+	}
+
+	return serviceerror.NewInvalidArgument("version history does not match")
```
Comment (wxing1292) [inline service/history/replication/executable_history_task.go:394]:
```go
lowestCommonAncestor, err := versionhistory.FindLCAVersionHistoryItem(e.versionHistoryItems, incomingHistoryItems)
if err != nil {
    return err
}

localLastItem, err := versionhistory.GetLastVersionHistoryItem(e.versionHistoryItems)
if err != nil {
    return err
}
incomingLastItem, err := versionhistory.GetLastVersionHistoryItem(incomingHistoryItems)
if err != nil {
    return err
}

if lowestCommonAncestor.Equal(localLastItem) || lowestCommonAncestor.Equal(incomingLastItem) {
    return nil
}
return serviceerror.NewInvalidArgument("version history does not match")
```

File: service/history/replication/executable_history_task.go [hunk 5]
```diff
@@ -297,3 +302,145 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
...
+
+	incomingHistoryTask, err := e.validateIncomingBatchTask(incomingTask)
+	if err != nil {
+		return nil, err
+	}
+
+	longerVersionHistory := e.versionHistoryItems
+	if len(incomingHistoryTask.versionHistoryItems) > len(longerVersionHistory) {
+		longerVersionHistory = incomingHistoryTask.versionHistoryItems
+	}
```
Comment (wxing1292) [inline service/history/replication/executable_history_task.go:325]: you should compare the event ID, not length

File: service/history/replication/executable_history_task.go [hunk 6]
```diff
@@ -297,3 +302,145 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
 	}
 	return events, newRunEvents, err
 }
+
+func (e *ExecutableHistoryTask) BatchWith(incomingTask BatchableTask) (_ TrackableExecutableTask, retErr error) {
+	if !e.wantToBatch {
```
Comment (yux0) [inline service/history/replication/executable_history_task.go:307]: so wanttobanch is one way，it doesn't support false to true I guess ？

File: service/history/replication/executable_history_task.go [hunk 7]
```diff
@@ -297,3 +302,145 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
...
+	if !(e.WorkflowKey.GetWorkflowID() == incomingWorkflowKey.GetWorkflowID() &&
+		e.WorkflowKey.GetNamespaceID() == incomingWorkflowKey.GetNamespaceID() &&
+		e.WorkflowKey.GetRunID() == incomingWorkflowKey.GetRunID()) {
+		return serviceerror.NewInvalidArgument("workflow key does not match")
+	}
+	return nil
+}
+
+func (e *ExecutableHistoryTask) checkBaseExecutionInfo(incomingTaskExecutionInfo *workflowpb.BaseExecutionInfo) error {
+	if e.baseExecutionInfo != nil && incomingTaskExecutionInfo != nil && !e.baseExecutionInfo.Equal(incomingTaskExecutionInfo) {
```
Comment (yux0) [inline service/history/replication/executable_history_task.go:407]: what about nil case?

File: service/history/replication/executable_history_task.go [hunk 8]
```diff
@@ -297,3 +302,145 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
...
+func (e *ExecutableHistoryTask) checkEvents(incomingEvents []*historypb.HistoryEvent, incomingNewRunEvents []*historypb.HistoryEvent) error {
+	if len(incomingEvents) == 0 {
+		return serviceerror.NewInvalidArgument("incoming task is empty")
+	}
+	currentEvents, currentNewRunEvents, err := e.getDeserializedEvents()
+	if err != nil {
+		return err
+	}
+
+	if currentNewRunEvents != nil && incomingNewRunEvents != nil {
```
Comment (yux0) [inline service/history/replication/executable_history_task.go:423]: If current new run !=nil, are we expected the task is not batchable?

File: service/history/replication/executable_history_task.go (hunk 6 above)
Comment (xwduan) [reply service/history/replication/executable_history_task.go:307]: Yes you are right.

File: service/history/replication/executable_history_task.go (hunk 7 above)
Comment (xwduan) [reply service/history/replication/executable_history_task.go:407]: BaseExecutionInfo is only for a certain usecase and it is either nil or equal. @wxing1292 could you confirm this?

File: service/history/replication/executable_history_task.go (hunk 8 above)
Comment (xwduan) [reply service/history/replication/executable_history_task.go:423]: Yeah, we should not expect the task is batchable because `currentNewRun != nil` indicating the current task is the last task. I will change the logic safer.

File: service/history/replication/executable_history_task.go (hunk 3 above)
Comment (xwduan) [reply service/history/replication/executable_history_task.go:306]: Sure. Will change.

File: service/history/replication/sequential_batch_queue.go (hunk 2 above)
Comment (xwduan) [reply service/history/replication/sequential_batch_queue.go:43]: Yes. Will remove.

File: common/persistence/versionhistory/version_history.go [hunk 9]
```diff
@@ -129,29 +129,54 @@ func ContainsVersionHistoryItem(v *historyspb.VersionHistory, item *historyspb.V
 
 // FindLCAVersionHistoryItem returns the lowest common ancestor VersionHistoryItem.
 func FindLCAVersionHistoryItem(v *historyspb.VersionHistory, remote *historyspb.VersionHistory) (*historyspb.VersionHistoryItem, error) {
-	localIndex := len(v.Items) - 1
-	remoteIndex := len(remote.Items) - 1
+	return findLCAVersionHistoryItemFromItemSlice(v.Items, remote.Items)
+}
+
+func findLCAVersionHistoryItemFromItemSlice(a []*historyspb.VersionHistoryItem, b []*historyspb.VersionHistoryItem) (*historyspb.VersionHistoryItem, error) {
```
Comment (wxing1292) [inline common/persistence/versionhistory/version_history.go:135]: what is a and b? can we use something which is meaningful?

File: common/persistence/versionhistory/version_history.go [hunk 10]
```diff
@@ -129,29 +129,54 @@ func ContainsVersionHistoryItem(v *historyspb.VersionHistory, item *historyspb.V
...
+			// aVersionItem.Version < bVersionItem.Version
+			bIndex--
 		}
 	}
 
 	return nil, serviceerror.NewInternal("version history is malformed. No joint point found.")
 }
 
+// IsSameBranch checks if two version history items are in the same branch
+func IsVersionItemsInSameBranch(a []*historyspb.VersionHistoryItem, b []*historyspb.VersionHistoryItem) bool {
```
Comment (wxing1292) [inline common/persistence/versionhistory/version_history.go:160]: this function should take in `*historyspb.VersionHistory` not `[]*historyspb.VersionHistoryItem`

File: common/tasks/sequential_scheduler.go [hunk 11]
```diff
@@ -267,6 +267,9 @@ func (s *SequentialScheduler[T]) processTaskQueue(
 			s.queueChan <- queue
 			return
 		default:
+			// sequential_batch_queue individual task handler is depending on following logic executed in synchronize:
+			// removeTaskFromQueue -> executeTask -> handleResult(Ack/Nack) -> removeQueue if Empty
+			// For the following reason: If batchedTask Nacked, the individual tasks in the batch will be put back to the queue
```
Comment (wxing1292) [inline common/tasks/sequential_scheduler.go:272]: i would reword the ^ to 

// NOTE: implicit assumption
// 1. a queue is owned by a coroutine
// 2. a coroutine will remove a task from its queue then execute the task; this coroutine will ack / nack / reschedule the task at the end
// 3. queue will be deleted once queue is empty
// 
// for batched tasks, if task is state
// ack: behavior is same as normal task
// nack: batched task will be broken into original tasks, and synchronously added to queue (so queue is not empty)
// reschedule: behavior is same as normal task

File: service/history/replication/batchable_task.go [hunk 12]
```diff
@@ -109,7 +111,10 @@ func (w *batchedTask) IsRetryableError(err error) bool {
 }
 
 func (w *batchedTask) RetryPolicy() backoff.RetryPolicy {
-	return w.batchedTask.RetryPolicy()
+	if len(w.individualTasks) == 1 {
```
Comment (wxing1292) [inline service/history/replication/batchable_task.go:114]: lock?

File: service/history/replication/batchable_task.go [hunk 13]
```diff
@@ -158,6 +163,9 @@ func (w *batchedTask) handleIndividualTasks() {
 }
 
 func (w *batchedTask) AddTask(incomingTask TrackableExecutableTask) bool {
+	if w.state != batchStateOpen {
+		return false
+	}
```
Comment (wxing1292) [inline service/history/replication/batchable_task.go:168]: lock?

File: service/history/replication/executable_history_task.go [hunk 14]
```diff
@@ -96,6 +101,7 @@ func NewExecutableHistoryTask(
 		versionHistoryItems: task.VersionHistoryItems,
 		eventsBlob:          task.GetEvents(),
 		newRunEventsBlob:    task.GetNewRunEvents(),
+		wantToBatch:         true,
```
Comment (wxing1292) [inline service/history/replication/executable_history_task.go:104]: batchable?

File: service/history/replication/executable_history_task.go [hunk 15]
```diff
@@ -299,3 +305,134 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
 	}
 	return events, newRunEvents, err
 }
+
+func (e *ExecutableHistoryTask) BatchWith(incomingTask BatchableTask) (TrackableExecutableTask, bool) {
+	if !e.wantToBatch {
```
Comment (wxing1292) [inline service/history/replication/executable_history_task.go:310]: locking?

File: common/persistence/versionhistory/version_history.go (hunk 10 above)
Comment (xwduan) [reply common/persistence/versionhistory/version_history.go:160]: Sure. Will add another method which take in *historyspb.VersionHistory.

File: service/history/replication/batchable_task.go (hunk 12 above)
Comment (xwduan) [reply service/history/replication/batchable_task.go:114]: sure.

File: service/history/replication/batchable_task.go (hunk 13 above)
Comment (xwduan) [reply service/history/replication/batchable_task.go:168]: This is a minor optimization. batchStateOpen is a one way switch. It can only change from `true` to `false`. So if batchStateOpen is `false`,  we can safely return false without locking.

File: service/history/replication/executable_history_task.go (hunk 15 above)
Comment (xwduan) [reply service/history/replication/executable_history_task.go:310]: Same as `batchStateOpen` in above, `wantToBatch` is also a one way switch.

File: service/history/replication/sequential_batch_queue.go (hunk 2 above)
Comment (xwduan) [reply service/history/replication/sequential_batch_queue.go:43]: Actually I decided to leave this var here to keep the interface consistent and as a reminder that the handler should be passed in. We are not using it now because we are taking a shortcut.

File: common/persistence/versionhistory/version_history.go (hunk 9 above)
Comment (wxing1292) [reply common/persistence/versionhistory/version_history.go:135]: can we use something more meaningful?

File: common/persistence/versionhistory/version_history.go (hunk 10 above)
Comment (wxing1292) [reply common/persistence/versionhistory/version_history.go:160]: >Sure. Will add another method which take in *historyspb.VersionHistory.

no this is not what i am asking
version contains 2 things
1. compressed info about event ID & version
2. persistence branch token
persistence branch token does not matter in this case (i.e. no divergence between 2 version histories)
 
```go
func IsVersionItemsInSameBranch(
    thisVersionHistory *historyspb.VersionHistory, 
    thatVersionHistory *historyspb.VersionHistory,
) bool {
    // business logic
}
```

File: service/history/replication/batchable_task.go (hunk 13 above)
Comment (wxing1292) [reply service/history/replication/batchable_task.go:168]: this kind of "optimization" makes no sense

1. prove that w.state can be safely accessed and the value will not break the logic here
2. this task is not going to be accessed by many goroutine, so how much nanoseconds will be saved? does this worth the complexity & potential data corruption?
3. use sync.Mutex or atomics or go chan from golang, do not rely on implicit assumption of any platform unless you write a ton of tests validating your impl is correct

File: service/history/replication/executable_history_task.go [hunk 16]
```diff
@@ -299,3 +305,134 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
...
+
+	currentEvents, currentNewRunEvents, _ := e.getDeserializedEvents()
+	incomingEvents, incomingNewRunEvents, _ := incomingHistoryTask.getDeserializedEvents()
+
+	return &ExecutableHistoryTask{
+		ProcessToolBox:      e.ProcessToolBox,
+		WorkflowKey:         e.WorkflowKey,
+		ExecutableTask:      e.ExecutableTask,
+		baseExecutionInfo:   e.baseExecutionInfo,
+		versionHistoryItems: e.getLongerVersionHistory(e.versionHistoryItems, incomingHistoryTask.versionHistoryItems),
```
Comment (wxing1292) [inline service/history/replication/executable_history_task.go:334]: getLongerVersionHistory -> latestVersionHistory

File: service/history/replication/executable_history_task.go [hunk 17]
```diff
@@ -299,3 +305,134 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
...
+	return serviceerror.NewInvalidArgument("version history does not match")
+}
+
+func (e *ExecutableHistoryTask) checkWorkflowKey(incomingWorkflowKey definition.WorkflowKey) error {
+	if !(e.WorkflowKey.GetWorkflowID() == incomingWorkflowKey.GetWorkflowID() &&
+		e.WorkflowKey.GetNamespaceID() == incomingWorkflowKey.GetNamespaceID() &&
+		e.WorkflowKey.GetRunID() == incomingWorkflowKey.GetRunID()) {
+		return serviceerror.NewInvalidArgument("workflow key does not match")
+	}
+	return nil
```
Comment (wxing1292) [inline service/history/replication/executable_history_task.go:395]: e.WorkflowKey != incomingWorkflowKey

File: service/history/replication/executable_history_task.go (hunk 7 above)
Comment (wxing1292) [reply service/history/replication/executable_history_task.go:399]: should also check if 

```go
if e.baseExecutionInfo == nil && incomingTaskExecutionInfo == nil {
    return nil
} else if e.baseExecutionInfo != nil && incomingTaskExecutionInfo != nil {
    if !e.baseExecutionInfo.Equal(incomingTaskExecutionInfo) {
        return err
    }
    return nil
} else {
    return err
}
````

File: service/history/replication/executable_history_task.go [hunk 18]
```diff
@@ -299,3 +305,134 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
...
+	}
+	if currentLastEvent.EventId+1 != incomingFirstEvent.EventId {
+		return serviceerror.NewInvalidArgument("events id is not consecutive")
+	}
+
+	return nil
+}
+
+func (e *ExecutableHistoryTask) CanBatch() bool {
+	return e.wantToBatch
```
Comment (wxing1292) [inline service/history/replication/executable_history_task.go:431]: lock?

Comment (wxing1292) [reply service/history/replication/executable_history_task.go:431]: if this function assumes that lock is already acquired, then rename to `canBatchLocked`

File: common/persistence/versionhistory/version_history.go (hunk 10 above)
Comment (xwduan) [reply common/persistence/versionhistory/version_history.go:160]: Yeah, I understand. The problem is that `ExecutableHistoryTask` and 'replicationspb.HistoryTaskAttributes' only has `[]*historyspb.VersionHistoryItem`. Shall we build a new VersionHistory object with empty branch token and then call this method?

**Review by wxing1292 — APPROVED** [review]

File: service/history/replication/executable_history_task.go [hunk 19]
```diff
@@ -299,3 +305,155 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
...
+			newRunEvents: append(currentNewRunEvents, incomingNewRunEvents...),
+			err:          nil,
+		},
+		batchable: true,
+	}, true
+}
+
+func (e *ExecutableHistoryTask) getFresherVersionHistoryItems(versionHistoryItemsA []*historyspb.VersionHistoryItem, versionHistoryItemsB []*historyspb.VersionHistoryItem) []*historyspb.VersionHistoryItem {
+	fresherVersionHistoryItems := versionHistoryItemsA
+	if versionHistoryItemsA[len(versionHistoryItemsA)-1].GetEventId() < versionHistoryItemsB[len(versionHistoryItemsB)-1].GetEventId() {
```
Comment (yux0) [inline service/history/replication/executable_history_task.go:346]: The two version history will have the same version?

Comment (yux0) [reply service/history/replication/executable_history_task.go:346]: I see checkVersionHistoryItem

**Review by yux0 — APPROVED** [review]
